tlog-witness.md

```diff
@@ -0,0 +1,226 @@
+the witness MUST respond with a "400 Bad Request" HTTP status code.
+
+The witness MUST check that the previous size matches the size of the latest
+checkpoint it cosigned for the checkpoint's origin (or zero if it never cosigned
+a checkpoint for that origin). If it doesn't match, the witness MUST respond
+with a "409 Conflict" HTTP status code. The response body MUST consist of the
+tree size of the latest cosigned checkpoint in decimal, followed by a newline
+(U+000A). The response MUST have a `Content-Type` of `text/x.tlog.size`.
+
+If a client doesn't have information on the latest cosigned checkpoint, it MAY
+initially make a request with a previous size of zero to obtain it.
+
+The consistency proof lines MUST encode a Merkle Consistency Proof from the
+previous size to the checkpoint size according to [RFC 6962][], Section 2.1.2. The
+proof MUST be empty only if the previous size is zero or if it matches the size
+in the checkpoint. If the Merkle Consistency Proof doesn't verify, the witness
+MUST respond with a "422 Unprocessable Entity" HTTP status code.
+
+If the origin is known, and the signature is valid, the witness SHOULD log the
+request even if the consistency proof doesn't verify (but MUST NOT cosign the
+checkpoint) as it might be proof of log misbehavior.
+
+If all the checks above pass, the witness MUST update its record of the latest
+cosigned checkpoint and respond with a "200 Success" HTTP status code. The
+response body MUST be a sequence of one or more [note][] signature lines, each
+starting with the `—` character (U+2014) and ending with a newline character
+(U+000A). The signatures MUST be [cosignatures][] from the witness key(s) on the
+checkpoint.
+
+    — witness.example/w1 CMp+6LWBU0anHGH5aNDTJkH/gj79sG+T6+iP2ThYN5krrDJbR1HDnucjL39QsZTSvVjyQLrdk3DXDqI5G2HgLatVs0pWh6Up69HVOw==
+    — witness.example/w1 I7rEps0pvK2UqkS2gSpVUDhrhVtQV9lgF6pRrWAvjJHjyWpW7VcE3SiOlVlbQNt64vWhO+DlkL0+UfzuOBMh9ChdMkP1vi/lCAsmlw==
+    — witness.example/w2 AWui8Sk55XjYLOijihBjhqEH6nS1ndDymE0a+6idX7pLcnoB+dhnz0854aLZgrrKbYKA7nC3HNJhm/kWl7oJlqU3rXXvpysAdyP3wQ==
+
+The client MUST ignore any cosignatures from unknown keys. To parse the
+response, the client MAY concatenate it to the checkpoint, and use a [note][]
+verification function configured with the witness keys it trusts. If that call
+succeeds, it can move the valid signatures to its own view of the checkpoint.
+
+The witness MUST persist the new checkpoint before responding. Note that
+checking the previous size against the latest checkpoint and persisting the new
+checkpoint must be performed atomically, otherwise the following race can occur:
+
+1. Request A with size N is checked for consistency.
+2. Request B with size N+K is checked for consistency.
+3. The stored size is updated to N+K for request B.
+4. A cosignature for N+K is returned to request B.
+5. The stored size is updated to N for request A, *rolling back K leaves*.
+6. A cosignature for N is returned to request A.
```

Comment:
This explanation is great, but borderline appendix material.  I.e., "Must be atomic to avoid race, see appendix" would maybe be neater.

Reply:
I tried but it came out awkward, because there are no other appendixes, and because it ended up being just the numbered list.